lib/spack/spack/test/packaging.py

```diff
@@ -124,69 +126,56 @@ def test_buildcache(mock_archive, tmpdir):
     buildcache.setup_parser(parser)
 
     # Create a private key to sign package with if gpg2 available
+    variants = {}
+    variants["jobs"] = map(str, [1, 2])
+    variants["relative"] = [False, True]
+    variants["signed"] = [False]
     if has_gnupg2():
-        spack.util.gpg.Gpg.create(name='test key 1', expires='0',
-                                  email='spack@googlegroups.com',
-                                  comment='Spack test key')
-        # Create build cache with signing
-        args = parser.parse_args(['create', '-d', mirror_path, str(spec)])
-        buildcache.buildcache(parser, args)
-
-        # Uninstall the package
-        pkg.do_uninstall(force=True)
-
-        # test overwrite install
-        args = parser.parse_args(['install', '-f', str(pkghash)])
-        buildcache.buildcache(parser, args)
-
-        # create build cache with relative path and signing
-        args = parser.parse_args(
-            ['create', '-d', mirror_path, '-f', '-r', str(spec)])
-        buildcache.buildcache(parser, args)
-
-        # Uninstall the package
-        pkg.do_uninstall(force=True)
+        variants["signed"].append(True)
+        key_created = False
+
+    for jobs, relative, signed in it.product(
+            *(variants[v] for v in "jobs relative signed".split())):
+        shutil.rmtree(mirror_path)
+        stage.destroy()
+        spack.mirror.create(
+            mirror_path, specs=[], no_checksum=True
+        )
+        stage.create()
 
-        # install build cache with verification
-        args = parser.parse_args(['install', str(spec)])
-        buildcache.install_tarball(spec, args)
-
-        # test overwrite install
-        args = parser.parse_args(['install', '-f', str(pkghash)])
-        buildcache.buildcache(parser, args)
-
-    else:
-        # create build cache without signing
-        args = parser.parse_args(
-            ['create', '-d', mirror_path, '-u', str(spec)])
-        buildcache.buildcache(parser, args)
-
-        # Uninstall the package
-        pkg.do_uninstall(force=True)
-
-        # install build cache without verification
-        args = parser.parse_args(['install', '-u', str(spec)])
-        buildcache.install_tarball(spec, args)
-
-        # test overwrite install without verification
-        args = parser.parse_args(['install', '-f', '-u', str(pkghash)])
-        buildcache.buildcache(parser, args)
-
-        # create build cache with relative path
-        args = parser.parse_args(
-            ['create', '-d', mirror_path, '-f', '-r', '-u', str(pkghash)])
-        buildcache.buildcache(parser, args)
+        # Create build cache with signing
+        args = ['create', '-d', mirror_path, "-j", jobs, str(spec)]
+        if not signed:
+            args.insert(-1, "-u")
+        elif not key_created:
```

Comment:
I did not touch the gpg-key-handling with my PR in any way - this could be some side-effects from [rebasing](https://github.com/electronicvisions/spack/commit/837974be138a08dab3e712734d1866a28dcc1f3) - for the rework I'll again rebase [our working solution](https://github.com/electronicvisions/spack/commit/837974be138a08dab3e712734d1866a28dcc1f3) on to the current `develop` branch.